### PR TITLE
Support conditionally depending on a provider

### LINKED ISSUE #652: A couple of null safety tweaks
Hi Remi,

We're looking at rolling the latest `package:provider` into google and hit a couple of blocking issues.

We have also taken the time to investigate and have a proposed solution ;)

Both issues only show up with mixed null safe and legacy code.

1. Legacy inherited providers can't contain nulls any more. This is because of the use of `_value is! T` to check for failed initialization. Legacy code can't supply an explicitly nullable `T`, so `null` is never `T`. Proposed fix: use an additional `bool` to track whether initialization succeeded: https://github.com/davidmorgan/provider/commit/1d6ee87cd48da981e5fb9302a831ec7b56b9bcd0

2. Legacy code might want to inject a `Provider<T?>`, because legacy code is used to handling nulls. But, legacy code can't write `T?` so it can't supply the right key--it can only ever request non-nullable providers. Proposed fix: when running in unsound (mixed) mode, look for `Provider<T?>` if `Provider<T>` is not found: https://github.com/davidmorgan/provider/commit/1a55c485cba486480e86e686fde48878410c24fb

Both proposed fixes include test coverage; adding test coverage is a bit awkward because it requires tests that run in mixed mode, which needs a bit of care when running `flutter test`; I update `flutter_test.sh` to do that.

Please let me know if this seems a good direction, if so I will clean up the code and send PRs. Or, of course, happy to discuss :)

Thanks :)

### PR DESCRIPTION
**Pull request by rrousselGit**
Fixes #652

### CODE REVIEW

**Comment by rrousselGit** [comment]
cc @davidmorgan

File: lib/src/provider.dart [hunk 1]
```diff
@@ -314,26 +338,27 @@ If you want to expose a variable that can be anything, consider changing
...
         return false;
       });
     } else {
       inheritedElement = context.getElementForInheritedWidgetOfExactType<
-          _InheritedProviderScope<T>>() as _InheritedProviderScopeElement<T>?;
+          _InheritedProviderScope<T?>>() as _InheritedProviderScopeElement<T?>?;
     }
 
-    if (inheritedElement == null) {
+    if (inheritedElement == null && null is! T) {
```
Comment (davidmorgan) [inline lib/src/provider.dart:357]: This one too, I think this needs guarding with `_isSoundMode`.

**Review by davidmorgan — COMMENTED** [review]
Thanks!

Great, it does look like this approach works--the impact of `T?` is strictly speaking visible in the public API but it looks like a rarely used part, so impact is minimal :) i.e. it's effectively hidden behind `Provider`. (As you said!)

The sheer number of cases to cover for full test coverage is intimidating, in case it's useful, I added them all for basic InheritedProvider inject here (but didn't tidy up the code): https://github.com/rrousselGit/provider/commit/0afbd42d9689e7c9cf07767703d465a62ce366d5 and those all work as expected/desired.

In the `of` check you guard the `value is! T` check with  `_isSoundMode`, which is what lets mixed mode continue to pass around nulls as before, allowing an incremental migration--I've commented on other such checks which should probably be guarded in the same way, although I did not take the time to create test cases so I'm not 100% sure :)

Thanks for looking into this--please let me know how I can help with next steps :)

**Comment by davidmorgan** [comment]
@rrousselGit Anything I can do to help move this forward, please? Not sure if you wanted me to adjust my PR in this direction or if you were planning on taking/landing it. Thanks.

**Comment by davidmorgan** [comment]
Hi hi -- any update please? :) No specific urgency from my side, but we definitely need this at some point. Thanks.

**Comment by davidmorgan** [comment]
Any update please?... Mostly to check that you're not waiting for me to do something :) thanks.

**Comment by rrousselGit** [comment]
Oh I'm sorry, I didn't receive a notification.

I plan to finish and release this feature this week-end

**Comment by davidmorgan** [comment]
@rrousselGit No worries--sounds good--looking forward to it :) thanks!

**Comment by rrousselGit** [comment]
@davidmorgan this was published in a dev release: 6.0.0-dev

If everything is alright, I will publish a stable release in a week or two

**Comment by davidmorgan** [comment]
@rrousselGit Thanks, much appreciated--I was already running our regression tests with it, all green. I will additionally verify that this unblocks incremental migration as hoped, and get back to you by Wed. I'm being specific about Wednesday as I'll be on vacation for a week from Thursday :)

Thanks again.

**Comment by davidmorgan** [comment]
@rrousselGit Confirmed it works as planned :) thanks!

**Pull request merged**